### ISSUE TITLE
Suggestion for new "standard" modules

### ISSUE THREAD

**Issue by GWRon** [bmx-ng/bmx-ng#45]
After a short hint of Brucey I will start this issue here.

## New standard modules
Brucey wrote a big bunch of modules which fill a gap almost any Blitzmax developer will reach once in a (bigger) projects lifetime.
Default brl-modules (brl.mod and the 3rd party modules pub.mod) for example lack convenience functions to retrieve user directories - which was not required on ancient Windows OS but nowerdays is needed for proper functionality (write access to certain directories is limited now).

For now Brucey already adds "sdl.mod" as new default module collection - and I am pretty sure there are some other "small" modules which could get added to the list of default modules (maybe also call it  "must have").

Pay attention to something: a default module __should__ be not as big as eg. "libxml.mod" as it should get auto-imported if you do not use the `framework` command. They should provide some barebone functionality. So back to the "xml" example I would prefer some lightweight XML-lib (tinyxml or so) which does a good but not perfect job with adding 4-5KB to the executable, instead of 1MB.

## Naming scheme
while "brl.mod" initially stood for Blitz Research Labs it was suggested to interpret them as "Blitz Runtime Libraries" - which could allow to use brl.mod as "namespace" of the new default modules too.
I would not be against a "default.mod" or "base.mod" ... namespace. Something which each blitzmax-user can recognize as something which is a "provided by default" thingy.

## Structure
pub.mod is there to house the "wrappers" for 3rd party libraries. Maybe this one can get reused too? Brl.mod then contains types helping to use the pub.mod-modules (so in essence they provide containers and helper functions to work with the 3rd party libraries).
Talking again about the XML stuff this means "libxml.mod" would belong to "pub.mod" as it wraps the libxml code. Within brl.mod (or something similar) you would then package serialization stuff utilizing the libXML module. Most "game" blitzmax-developers _should_ prefer writing `SerializeObject(obj, filename)`. Similar to calling `LoadImage()` instead to directly communicating with FreeImage-code.

If you are interested to help - feel free to add your suggestions here (I might edit them to this message). Also discussion about structure, naming scheme etc are nice to have.

**Comment by GWRon** [comment]
I would for example suggest:

default:
- clipboard: https://github.com/maxmods/bah.mod/tree/master/clipboard.mod
- json support: https://github.com/maxmods/bah.mod/tree/master/jansson.mod (instead of libXML?)
- xml support: https://github.com/maxmods/bah.mod/tree/master/libxml.mod (or better use json?)
- xml lite support: none found yet but I think some "basic xml support" should be provided (no xlst, xpath, ... needed)
- serialization stuff: https://github.com/maxmods/bah.mod/tree/master/persistence.mod (or the json variant)
- stringbuilder: https://github.com/maxmods/bah.mod/tree/master/stringbuilder.mod (it's small but efficient!)
- user directories: https://github.com/maxmods/bah.mod/tree/master/volumes.mod

required by certain platforms: 
- sdl: https://github.com/bmx-ng/sdl.mod

non default ("without framework") but provided:
- box2D: https://github.com/maxmods/bah.mod/tree/master/box2d.mod


nice to have but not yet finished or "incorporated":
Col's / @davecamp's render2texture at https://github.com/davecamp/Render2Texture is something I would see provided as "basic functionality". It should be as simple as a "DrawImage()" to render to a texture. For now it needs some stuff to work properly with SDL (and so other platforms). Maybe col is interested in bringing the module in the perfect shape to get included as default module?

**Comment by HurryStarfish** [comment]
Yep, we are already "reusing" `BRL.mod` and `Pub.mod` and extending them with new modules (`BRL.Collections`, `Pub.xmmintrin`, ...). Renaming either of them is not an option, considering it would break pretty much literally all existing BlitzMax code out there. I think we are fine with these two scopes for the standard modules.

A list like this is a good thing to have. The standard libraries could indeed use some additions (as well as improvements in certain places).
The string builder module you're suggesting is already in `BRL.Stringbuilder`, so I think you can remove that from the list.
As an addition, I'd like to suggest some modules for non-trivial math datatypes, like big integers, complex numbers and vectors. (I tried writing a basic complex number and vector module some time ago; I ran into the problem of https://github.com/bmx-ng/bmx-ng/issues/19 though.)

**Comment by GWRon** [comment]
Yes vectors might be a good thing too. Could replace my basic approaches at GWRon/Dig.git .
Mapm.mod for scientific numbers is good stuff but I think it is not really a mandatory module. There is bigint.mod which I would classify similar.

**Comment by HurryStarfish** [comment]
Yeah, `BaH.BigInt` exists, but seems very barebones/unfinished. It could probably be extended though. (And eventually maybe moved into `Pub.mod`?)

**Comment by woollybah** [comment]
I've added bah.bsdnt, which is an NG-based big integer module. Bah.bigint was originally added as a dependency for another module. The library is not currently maintained, so there wasn't really much incentive to spend much time implementing it more fully.

**Comment by GWRon** [comment]
Any other input on what might be useful as standard module for a blitzmax user?
Dunno how much we are still targetting "game developers" instead of "tool and little game creators" ... depending on that we could add a bit of scientific stuff, console stuff (ncurse or so) ...
I would prefer to stay with "games", so concentrate on asset loading, ressource management, ...

**Comment by woollybah** [comment]
I've been working on a bunch of vector/matrix additions to brl.math.
They are all Structs and utilise method overloading where useful. So far I have SVec2, SVec3, SMat2, SMat3, SMat4 and SQuat.

**Comment by woollybah** [comment]
I'm also leaning towards having SDL as the primary backend for macOS now, given I don't really have the time to rewrite brl.glgraphics and friends.
I'd also like to insert bgfx into the mix as the low level processor between Max2D and SDL, which would allow us to target things like Metal and Vulkan indirectly - to that it basically wants something like sdl.gl2sdlmax2d to be ported to bgfx, rather than be GL specific. bgfx uses its own GLSL-type shaders, so in theory, it just wants the appropriate shaders implemented...

I've added the BlitzMax samples to bmx-ng, updated for NG, and using the SDL backend (because, why not).

**Comment by GWRon** [comment]
So Max2D should utilize bgfx.mod - and bgfx.mod uses sdl.mod/* ?

Think a bunch of "maxgui/wxmax"-things might break then (I vaguely remember to have had issues with sdl.mod and wxmax/my gui apps).

**Comment by woollybah** [comment]
I've just added Pub.mxml and BRL.xml modules.
The mxml library is much smaller, and therefore does not implement as many features as libxml (like compression, namespaces, etc), but provides a reasonable base from which to do xml. Users wanting more, can "upgrade" to bah.libxml.
The API closely resembles the BaH.libxml API in order to make transitioning from one to the other fairly trivial.

**Comment by GWRon** [comment]
Good job with mxml (will try soon - so prepare for issue reports :-)).

Using SDL for the context creation ... hmm, doesn't it open a can of worms? IF we really used SDL there, the whole thing should be based on SDL (timers etc) - with it's advantages and disadvantages.

But isn't there some cross-platform thingy for the context / window stuff too? I only know of GLFW but it is not for DX ...

**Comment by GWRon** [comment]
I am asking above as "SDL" consists of so many things ... and having "beasts" within the modules will sooner or later lead to either "duplicated functionality" or naming confusing or even more trickier problems.
Having a simple module/library for a limited purpose is more like the initial "module" approach of BlitzMax.
In other words: you should always think twice if you need multiple "sub folders" for modules (sdlimage, sdl, sdlaudio, ...) - even with bgfx you already get a load of modules justifying the use of a custom "main module name" ([b]gfx.mod).

So _if_ possible we should try to have some kind of "brl.mod/graphicswindow.mod" or so. It should be avoided to replace all the timer stuff (like sdl.mod does for now) - until of course this is all automatically handled and does not lead to issues putting of newcomers.

**Comment by GWRon** [comment]
"stringbuilder" could be added to a generic "brl.stringhelper" or "brl.string" module containing other functionality too.
Same as a "brl.math" might contain all the vector stuff and so on.

**Comment by GWRon** [comment]
Updated my list above with current state.

Anybody with other ideas for default modules?

**Comment by HurryStarfish** [comment]
A DateTime module maybe? For handling time stamps, dates, durations, time zones etc.
Considering that writing a library for that stuff is a *ludicrous* amount of effort, this is definitely not something we can make ourselves; but maybe there's a good open-source one somewhere that could be wrapped in a `Pub` module.

**Comment by GWRon** [comment]
So something like Chrono (C++):
https://en.cppreference.com/w/cpp/chrono 
?

A basic variant (but on top of the whole "boost" library) is bah.datetime:
https://github.com/maxmods/bah.mod/tree/master/datetime.mod

**Comment by woollybah** [comment]
"Proper" datetime libraries are large - on account of the many and varying locales to cater for. I'm not sure there are any "small" complete libraries for doing such things.

What else is useful? A Windows Registry module?
I have an enhanced brl.math module with Vec/Matrix/Quat implementations (driven by overloaded operators), if that's useful.

**Comment by GWRon** [comment]
Is there something speaking against joining that stuff into brl.math (except for a neglectable size-increase)? Another option is to call it brl.vectormath or so...in case it is a slightly less often used feature.

**Comment by HurryStarfish** [comment]
Similar to the XML thing we discussed, namespace pollution and collisions. `BRL.Math` contains basic functions like `Sqr`, but people might want to use their custom/different vector types (or none at all). Makes sense to have them separate.
The modules can be named `BRL.Vector`, `BRL.Matrix` etc.

**Comment by GWRon** [comment]
Regarding separation: you will have this issue with all "basic modules". People might have their own "TImage", "TSound", ...
The more "generic" or "basic" you name your type the more likely you will run into collisions.

Regarding "Brl.Vector" - this is a valid point as a vector is not just used for math ("BRL.VectorMath") but also for entity positions or similar stuff. So it makes sense to have such stuff in an individual module.


@ Standard module
Tweening might be useful (with support of the vectors inbuilt?). At least for casual games this is used very often.

**Comment by HurryStarfish** [comment]
`People might have their own "TImage", "TSound"`
Well yeah, in which case they're not gonna use `BRL.Max2D` and `BRL.Audio` which they're defined in. They're not packed together with something as essential as `Sqr`, `Sin` and `Floor`.

**Comment by GWRon** [comment]
You are right about sin/sqr etc.

**Comment by GWRon** [comment]
Is there a "sha/md5/blowfish/..." module already available? Maybe having such functionality (lightweight of course) could be interesting (eg to create file hashes to check for corruption/manipulation).

**Comment by GWRon** [comment]
As @woollybah added some new modules to brl.mod and things like text.mod now exist...we might rethink about what is still needed.

Wasnt boost.mod added too? In that case we could bring other bah.rational etc then too.
